pkg/controllers/federatedcluster/clusterjoin.go

```diff
@@ -80,232 +75,162 @@ const (
+			Status:  corev1.ConditionFalse,
+			Reason:  TokenNotObtainedReason,
+			Message: msg,
+		}, err
 	}
+
+	// 3. Get or create system namespace in the cluster, this will also tell us if the cluster is unjoinable
+
+	logger.Info(fmt.Sprintf("Create or get system namespace %s in cluster", fedSystemNamespace))
+	memberFedNamespace, err := clusterKubeClient.CoreV1().Namespaces().Get(ctx, fedSystemNamespace, metav1.GetOptions{ResourceVersion: "0"})
 	if err != nil {
-		return cluster, fmt.Errorf("failed to get cluster secret: %w", err)
+		if !apierrors.IsNotFound(err) {
+			msg := fmt.Sprintf("Failed to get namespace: %v", err.Error())
+			logger.Error(err, "Failed to get namespace")
+			eventRecorder.Eventf(
+				cluster,
+				corev1.EventTypeWarning, EventReasonJoinClusterError, msg,
+			)
+			return cluster, &fedcorev1a1.ClusterCondition{
+				Status:  corev1.ConditionFalse,
+				Reason:  TokenNotObtainedReason,
+				Message: msg,
+			}, err
+		}
+
+		// Ensure the value is nil and not garbage.
+		memberFedNamespace = nil
 	}
 
-	if err := util.PopulateAuthDetailsFromSecret(restConfig, cluster.Spec.Insecure, clusterSecret, false); err != nil {
+	if memberFedNamespace != nil && memberFedNamespace.Annotations[FederatedClusterUID] != string(cluster.UID) {
+		// ns exists and is not created by us - the cluster is managed by another control plane
+		msg := "Cluster is unjoinable (check if cluster is already joined to another federation)"
+		logger.Error(nil, msg, "UID", memberFedNamespace.Annotations[FederatedClusterUID], "clusterUID", string(cluster.UID))
+		cluster.Status.JoinPerformed = false
 		eventRecorder.Eventf(
 			cluster,
-			corev1.EventTypeWarning,
-			EventReasonHandleTerminatingClusterFailed,
-			"cluster %q secret %q is malformed: %v",
-			cluster.Name,
-			clusterSecretName,
-			err.Error(),
+			corev1.EventTypeWarning, EventReasonClusterUnjoinable, msg,
 		)
-		return cluster, fmt.Errorf("cluster secret malformed: %w", err)
+		return cluster, &fedcorev1a1.ClusterCondition{
+			Status:  corev1.ConditionFalse,
+			Reason:  ClusterUnjoinableReason,
+			Message: msg,
+		}, nil
 	}
 
-	clusterKubeClient, err := kubeclient.NewForConfig(restConfig)
-	if err != nil {
-		return cluster, fmt.Errorf("failed to create cluster kube clientset: %w", err)
-	}
-
-	// 3. Create or get system namespace in the cluster, this will also tell us if the cluster is unjoinable
-
-	logger.Info(fmt.Sprintf("Create or get system namespace %s in cluster", fedSystemNamespace))
-	memberFedNamespace, unjoinable, err := getOrCreateFedSystemNamespace(clusterKubeClient, fedSystemNamespace, string(cluster.UID))
-	if err != nil {
+	// Either the namespace doesn't exist or it is created by us.
+	// Clean-up on removal is required.
+	cluster.Status.JoinPerformed = true
+	// Update the condition to persist the flag before proceeding to ensure
+	// we don't lose the flag due to an update failure.
+	if cluster, err = client.CoreV1alpha1().FederatedClusters().UpdateStatus(
```

Comment:
Should we check if the original status already has `JoinPerformed` set to true to avoid unnecessary updates?